### ISSUE TITLE
ITC-3 reports occupancy (S0207) out of range

### ISSUE THREAD

**Issue by otterdahl** [rsmp-nordic/rsmp_validator#453]
In recent runs, the ITC-3 reports S0207 occupancy with very large values. However, according to the specification S0207 "occupancy" needs to be between 0 and 100.
```
[
        {
            "sCI": "S0207",
            "n": "occupancy",
            "q": "recent",
            "s": "201,240,238,11583606,11583606,11583606,11583606,11583606,11583606,11583606"
        }
    ]
```

**Comment by otterdahl** [comment]
The validator doesn't catch this issue. Perhaps the test case can be extended

**Comment by emiltin** [comment]
The limit is specified in out yaml source at:
https://github.com/rsmp-nordic/rsmp_schema/blob/main/schemas/tlc/1.2.1/sxl.yaml#L1017

However, it's not currently used when we convert the yaml to this JSON schema:
https://github.com/rsmp-nordic/rsmp_schema/blob/main/schemas/tlc/1.2.1/statuses/S0207.json#L65

The reason is that values are send as strings, making validation overly complex with json schema, see:
https://github.com/rsmp-nordic/rsmp_schema/issues/94

I think it would be best to catch this in the schema validation, rather than adding code to the validator tests - let's catch issues as soon as possible, we already have the limit defined in the sxl yaml.

**Comment by emiltin** [comment]
I'll extend the test to check that values are in the allowed range 0-100.

**Comment by emiltin** [comment]
#528 reports these illegal values this for the ITC-3.

**Comment by sveitech** [comment]
Please see: https://github.com/rsmp-nordic/rsmp_sxl_traffic_lights/issues/206

**Comment by sveitech** [comment]
I am saying that the result is undefined. It is not specified what should happen when you query this object. If you just issue a query, since when did the occupancy start measuring? Since the daemon started? Since 1970? e.t.c.

**Comment by emiltin** [comment]
Right. A simple solution would be to prohibit requests for these statuses.

But we could choose to define a meaningful use for requests and send on change. They could return the the latest single (not aggregated) measurement and its timestamp, i.e:

|  | S0001 | S0002 | S0003 |
| - | - | - | - |
| sOc/request | event | speed | presence |
| uRt (aggregation) | count | average speed | occupancy |

For S0003 and S0207, `start` would then be the timestamp of the latests measurement, and `occupancy` would be either 0 or 100 (or -1 if no measurements was recorded).

For S0206-8, you send a list of messurements separated by commas, one for each detector logic. With request on send on change, detectors that did not record anything could be set to -1, e.g. "-1,-1,-100" in case DL3 changdd, but 1 and 2 did not.

Generally, it would be good with a more structured approach to aggregation in relation to status updates. See this discussion related to a possible new SXL for sensors: https://github.com/rsmp-nordic/rsmp_sxl_sensors/issues/5

Occupancy can be seen as the average of presence (0 or 1) over time.

|  | count | speed | presence |
| - | - | - | - |
| type | pulse | bit | float |
| sum | count | | |
|average| | average speed | occupancy |

**Comment by SwarcoPalm** [comment]
Then they will all boil down to presence.

The measurement is updated (on ITC3) every 100ms, so you would get quite a lot of messages from onChange. Otherwise a standard interval should be defined, maybe like 15 mins. So that if you subscribe on change you will actually get a recurring message every 15 mins, and if you issue a status request you will get from the latest 15 min bin.

These would also not make sense to ask for without requesting the start time, but then we leave some small opening for the central system to shoot itself in the foot, maybe that is okay.

The balance here is between something that can be tested with the validator (where status request makes more sense than waiting for a long period), and the real world use-case which would pretty much always be recurring subscription.

**Comment by emiltin** [comment]
How would speed be reduced to presence?

Even though presence is measured every 100ms, I think send on change would only report a change when that measurement changed. This is based on the idea that with send on change, S0003 would report presense (0 or 100), not an average.

Agree with the balance, although I think sometimes it can be useful to be able to view live data.

**Comment by emiltin** [comment]
When you subscribe to a status, a status update is send immediately. Does this return the aggregated value for the latest full period?

**Comment by emiltin** [comment]
In PR #536, the tests has been changed to use a status subscription instead of a request. But as mentioned above, what occupancy value is the initial status update expected to contain?

**Comment by sveitech** [comment]
@emiltin We are aware of the issue. A solution will be made when we agree on how these objects should actually work.

**Comment by emiltin** [comment]
The test now uses a status subscription rather than a request. You question is what occupancy value should be send in the initial status update?

**Comment by emiltin** [comment]
General discussion 

> Alternatively, should we disable on-change subscriptions entirely for these objects and rule that they only work for interval-subscriptions (my preferred option as I think it is the least confusing).

Issue on this at  https://github.com/rsmp-nordic/rsmp_sxl_traffic_lights/issues/213

**Comment by SwarcoPalm** [comment]
I propose to close this issue considering the overall discussion is elsewhere and the ITC implementation is now updated to work in line with the subscription based approach.

**Comment by emiltin** [comment]
> I propose to close this issue considering the overall discussion is elsewhere and the ITC implementation is now updated to work in line with the subscription based approach.

ok
